### PR TITLE
fix wrong behavior of some PATCH endpoints

### LINKED ISSUE #44: Invalid PATCH behaviour on place, offer endpoints
## Current behavior
Making a PATCH request to `/place/3` while logged in as a user who owns place `id=3`, with the following body:
```json
{
  "address": "abcabcabc"
}
```
results in:
```json
{
  "status": "Forbidden",
  "code": 403,
  "message": "Cannot modify other place of another user as a regular user"
}
```

## Expected behavior
Server should PATCH the place with supplied data.

### PR DESCRIPTION
**Pull request by papaj-na-wrotkach**
should fix #44

### CODE REVIEW

File: server/src/controller/OfferController.hh [hunk 1]
```diff
@@ -200,7 +200,8 @@ class OfferController : public oatpp::web::server::api::ApiController
         OATPP_ASSERT_HTTP(
           auth_object->user_role == Role::Admin ||
             auth_object->user_id ==
-              place_service_.getOne(dto->place_id)->owner_id,
+              place_service_.getOne(offer_service_.getOne(id)->place_id)
+                ->owner_id,
```
Comment (papaj-na-wrotkach) [inline server/src/controller/OfferController.hh]: `dto->place_id` can be null as it's not required in PATCH method

File: server/src/controller/PlaceController.hh [hunk 2]
```diff
@@ -199,8 +199,8 @@ class PlaceController : public oatpp::web::server::api::ApiController
     {
         OATPP_ASSERT_HTTP(
           auth_object->user_role == Role::Admin ||
-            (dto->owner_id == auth_object->user_id &&
-             dto->owner_id == service_.getOne(id)->owner_id),
+            ((!dto->owner_id || dto->owner_id == auth_object->user_id) &&
+             auth_object->user_id == service_.getOne(id)->owner_id),
```
Comment (papaj-na-wrotkach) [inline server/src/controller/PlaceController.hh:203]: `dto->owner_id` can be null as it's not required in PATCH method (we should check `user_id` from `auth_object` as it cannot be null)

Comment (papaj-na-wrotkach) [reply server/src/controller/PlaceController.hh:203]: wait, I need to change it as it's wrong

Comment (papaj-na-wrotkach) [reply server/src/controller/PlaceController.hh:203]: changed (why ISO C++ forbids omitting middle operand in tenary T~T)

Comment (Xertes0) [reply server/src/controller/PlaceController.hh:203]: For me it still allows to change someone else's offer and parking place.

**Review by Xertes0 — CHANGES_REQUESTED** [review]
#44 was fixed, but now it allows to PATCH offers and places not owned by the authenticated user.

**Comment by Xertes0** [comment]
> #44 was fixed, but now it allows to PATCH offers and places not owned by the authenticated user.

Sorry, I think that was my problem. I can't replicate it again.

File: server/src/controller/PlaceController.hh (hunk 2 above)
Comment (papaj-na-wrotkach) [reply server/src/controller/PlaceController.hh:203]: Check again. It does not allow me to PATCH neither a place nor an offer which does not belong to authorized user

**Comment by papaj-na-wrotkach** [comment]
Can it be positively reviewed to merge into main?

**Comment by Xertes0** [comment]
> > #44 was fixed, but now it allows to PATCH offers and places not owned by the authenticated user.
> 
> Sorry, I think that was my problem. I can't replicate it again.

Okay, I replicated it.
It is possible to overwrite someone else's offer and parking place, by changing the id in PATCH request.
Example:
`PATCH /place/4` with place id=4 owned by the authenticated user, and place id=1 by someone else.
```json
{
  "id": 1,
}
```
This request will overwrite place id=1, no matter who owns it.

**Comment by papaj-na-wrotkach** [comment]
It's weird. DB operation should fail as `id` is unique...

**Comment by Xertes0** [comment]
> > > #44 was fixed, but now it allows to PATCH offers and places not owned by the authenticated user.
> > 
> > 
> > Sorry, I think that was my problem. I can't replicate it again.
> 
> Okay, I replicated it. It is possible to overwrite someone else's offer and parking place, by changing the id in PATCH request. Example: `PATCH /place/4` with place id=4 owned by the authenticated user, and place id=1 by someone else.
> 
> ```json
> {
>   "id": 1,
> }
> ```
> 
> This request will overwrite place id=1, no matter who owns it.

Another update:
It seems that it doesn't result in overwriting place id=1, but in changing the id to 1 (how it almost should), resulting in two places with the same id.

**Comment by papaj-na-wrotkach** [comment]
Ah, I know. I did use `PUT` endpoint's method to `PATCH` entity providing it with missing information from the DB

**Comment by papaj-na-wrotkach** [comment]
`PUT` endpoints method executes `REPLACE` operation on the DB

File: server/src/service/OfferService.cc [hunk 3]
```diff
@@ -137,19 +136,42 @@ OfferService::putOne(Object<OfferDto> const& dto)
 Object<OfferDto>
 OfferService::patchOne(UInt64 const& id, Object<OfferDto> const& dto)
 {
-    auto existing = this->getOne(id);
+    validateDateHTTP(dto->date_from, dto->date_to);
```
Comment (Xertes0) [inline server/src/service/OfferService.cc]: `dto->date_from` and `dto->date_to` might not be present in request's body.

**Review by Xertes0 — CHANGES_REQUESTED** [review]

File: server/src/service/OfferService.cc (hunk 3 above)
Comment (papaj-na-wrotkach) [reply server/src/service/OfferService.cc]: fixed?

Comment (Xertes0) [reply server/src/service/OfferService.cc]: > fixed?

It doesn't build.

Comment (papaj-na-wrotkach) [reply server/src/service/OfferService.cc]: accidentally deleted one `{`

**Comment by Xertes0** [comment]
Now when you want to change place's id, to an already taken one, the server responds with 500, shouldn't it be 403 - Forbidden?

File: server/src/service/OfferService.cc (hunk 3 above)
Comment (papaj-na-wrotkach) [reply server/src/service/OfferService.cc]: now it builds

**Comment by papaj-na-wrotkach** [comment]
imo 409 better

**Comment by Xertes0** [comment]
Didn't change anything for me.
Query: `PATCH /place/10` with place id=4 owned by the authenticated user
```json
{
  "id": 1
}
```
Result:
```json
{
  "status": "Internal Server Error",
  "code": 500,
  "message": "UNIQUE constraint failed: place.id"
}
```

**Comment by Xertes0** [comment]
As a side note:
I managed to reproduce #50 with GCC and Clang, by simply attaching a debugger, and stopping at a break point.

**Comment by Xertes0** [comment]
Okay, this seems to work, but now I noticed, we can change our offer's `place_id` to someone else's.

**Comment by papaj-na-wrotkach** [comment]
As we need more and more validation in this project I was thinking about how to do it better...

If we were to develop this project further, I would create a static `DataValidator` class that would store all the validation methods, but I don't think we need it as we are finishing development on this project. What is your opinion @Xertes0 ?

**Comment by Xertes0** [comment]
> If we were to develop this project further, I would create a static `DataValidator` class that would store all the validation methods, but I don't think we need it as we are finishing development on this project. What is your opinion @Xertes0 ?

Right now, the only things that require 'complex' validation are emails and dates.

Dates are supplied only in offers, and the function is only ~5 lines long, so I think it's fine to keep it in `OfferService.cc`.

Emails are used in both user controller and auth controller. That is the reason it's been split off to a separate file.

As things are now, I think it's better to, as you said, don't change anything, because we are about to finish this project anyway.

**Comment by papaj-na-wrotkach** [comment]
> Okay, this seems to work, but now I noticed, we can change our offer's `place_id` to someone else's.

Yes, it's fixed now. I changed my previous code to be similar to new code

**Comment by Xertes0** [comment]
> > Okay, this seems to work, but now I noticed, we can change our offer's `place_id` to someone else's.
> 
> Yes, it's fixed now. I changed my previous code to be similar to new code

Now, I can't change `place_id` to my own place.

**Comment by papaj-na-wrotkach** [comment]
I did sth really stupid: `dto->place_id == auth_object->user_id`

**Comment by papaj-na-wrotkach** [comment]
now finally (maybe) it is (should be) working

**Comment by papaj-na-wrotkach** [comment]
I checked for myself and confirmed - it is fixed

**Comment by Xertes0** [comment]
Now I'm able to PATCH an offer with `place_id` owned by some else, to a place owned by me.

**Comment by papaj-na-wrotkach** [comment]
AAAAAAAA
A
A
AAA
AAAAH
 
 
 
 
I forgor
 
 
I will change it
 
 
 
 
I am sorry that you have to check it so many times
 
 
I think it's fixed now
 
 
 
Is it?

**Review by Xertes0 — APPROVED** [review]
Looks good. 👍🏻

**Pull request merged**